### PR TITLE
Validator: implement ```eachElementOf(v).satisfies(predicate)```

### PR DESCRIPTION
**Pull request by JoelGunawan**
Example usage:

```cpp
public:
    void Constraints() {
        CONS(eachElementOf(A).satisfies(TestSpec::isOdd));
    }

private:
    static bool isOdd(int val) {
        return val % 2 == 1;
    }

```

### CODE REVIEW

**Comment by JoelGunawan** [comment]
Currently eachElementOf(v).satisfies(predicate) only supports scalar types.

File: include/tcframe/validator/core.hpp [hunk 1]
```diff
@@ -49,6 +51,15 @@ struct VectorElementValidator {
         }
         return true;
     }
+    
+    bool satisfies(function<bool(T)> predicate) {
+        for(T v: val) {
```
Comment (fushar) [inline include/tcframe/validator/core.hpp:56]:
```suggestion
        for (T v : val) {
```

File: include/tcframe/validator/core.hpp [hunk 2]
```diff
@@ -49,6 +51,15 @@ struct VectorElementValidator {
...
+            if(!predicate(v)) {
```
Comment (fushar) [inline include/tcframe/validator/core.hpp:57]:
```suggestion
            if (!predicate(v)) {
```

File: include/tcframe/validator/core.hpp [hunk 3]
```diff
@@ -150,6 +161,15 @@ struct MatrixElementValidator {
         }
         return true;
     }
+
+    bool satisfies(function<bool(T)> predicate) {
+        for(const vector<T>& v : val) {
```
Comment (fushar) [inline include/tcframe/validator/core.hpp:166]:
```suggestion
        for (const vector<T>& v : val) {
```

File: include/tcframe/validator/core.hpp [hunk 4]
```diff
@@ -150,6 +161,15 @@ struct MatrixElementValidator {
...
+            if(!eachElementOf(v).satisfies(predicate)) {
```
Comment (fushar) [inline include/tcframe/validator/core.hpp:167]:
```suggestion
            if (!eachElementOf(v).satisfies(predicate)) {
```

**Review by fushar — APPROVED** [review]

**Pull request merged**